src/help/help.cpp

```diff
@@ -87,16 +87,14 @@ void show_unit_description(const unit_type &t)
 		help::show_unit_help(t.id(), t.show_variations_in_help(), hide_help);
 }
 
-help_manager::help_manager(const game_config_view *cfg) //, gamemap *_map)
+static game_config_view dummy_view;
+help_manager::help_manager(const game_config_view *cfg)
+	: guard(game_cfg, cfg == nullptr ? &dummy_view : cfg)
```

Comment:
What can this `guard` do, other than re-introducing a dangling pointer at a later point in time?

The old code was easier to read, and I think it would behave the same unless a dangling pointer had already been created before this method was called. If that dangling pointer does exist, then this just makes debugging harder.

Reply:
This guard makes help managers nestable. If you had two help managers active simultaneously, with the old code the pointer would be nulled out when the inner one exits, meaning an attempt to show help after that could cause a crash. With this, it'll be nulled out when the outer one exits, and if by some chance there were two different game configs (I don't think there ever would be though), the outer one would be restored when the inner help manager exits.

Reply:
Replied to in the other PR, I still consider this an issue in this PR.

Reply:
In what situation would there be two help managers active?

Reply:
If you call `gui.show_help()`, because in order to ensure the help exists no matter where it's called from, it needs to create a help manager.

Reply:
@CelticMinstrel, will you now remove this?

Reply:
Just realised that nesting the help like that would be a UX regression.

The help browser already has a back button; from a UX perspective, opening a second help page already does nest. Once it's nested there are two buttons available - one to unnest by one level, and another to immediately exit the help browser regardless of how many levels are nested.

The next question is what would trigger a second help manager to open, and again that would be a UX regression if it was anything other than the UI that the user was interacting with. From a UX perspective, I think the only way `gui.show_help()` should be able to open a second help page would be if we added a feature that allowed Lua or ActionWML for interactive help pages; and even in that case, the correct UX would be to open a new page in the existing help browser, so that the back button works.

Reply:
Your comment makes no sense whatsoever. The help manager has nothing to do with UX; all it does is make sure help has been parsed and generated topics have been generated.

Additionally, I tested it and there doesn't seem to be any weird behaviour with the back button. It works as expected.

Some parts of your comment would be relevant if a help topic were able to contain links or buttons that execute Lua code which could theoretically call `gui.show_help` again, but since that's not possible right now, I think it's something we can worry about if it becomes possible in the future.

Reply:
You said several times "your comment makes no sense" in the RAII thread. Please learn from that thread.

@Pentarctagon does it make sense to you?

Reply:
It sounds like this change essentially results in a no-op, since there's currently no way to have two help managers existing at the same time right now.  In which case I would say that this specific change should be removed until/unless it becomes possible in the future.

Reply:
It's not a no-op, because this commit _creates_ a situation where there can be two help managers existing at the same time. Now, I _could_ go back to exposing help only in the game kernel, making this become unnecessary; but I think it should be done anyway, as the previous help manager seems like a crash waiting to happen.

From my perspective, there's no reason _not_ to do this change. It has no impact on _anything_, except making the rest of the commit safe and crash-free. None of the arguments you made against using `guard_value` in the other thread apply here, either - there, I replaced one resource management idiom with another, and you were reasonably concerned that there might be unforeseen consequences from that change; but here, the code was _already_ using the `guard_value` idiom, just in a non-reentrant way.

There is technically a third option, which is to remove the help manager altogether and make all the help code directly access the game config whenever help is shown. I have no idea what unforeseen consequences that might have, though.

> You said several times "your comment makes no sense" in the RAII thread.

I said it doesn't make sense because it doesn't make sense. Your comment was about things that have no relationship whatsoever with this change. The help manager has nothing to do with any user interface code; it's just a scope guard, nothing more.

> Please learn from that thread.

That thread was a completely different situation. We apparently had a different definition of what RAII is, so we ended up talking past each other for awhile until that discrepancy was brought to light. That's not what happened here. Anything I might learn from that thread, while likely useful in some capacity, doesn't apply here.

-----

Just to be clear, the following two requests are wholly incompatible. I can do one of them, but not both.

1. Expose `gui.show_help` in lua_kernel_base.
2. Revert this change to the help manager.

Reply:
Okay, I misunderstood how long the lifecycle of the main help_manager is. I'm now re-reviewing this.